pkg/dns/aws/filewatcher.go

```diff
@@ -0,0 +1,76 @@
+package aws
+
+import (
+	"bytes"
+	"io/ioutil"
+	"time"
+
+	certsutil "github.com/openshift/cluster-ingress-operator/pkg/util/certs"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+// startWatcher starts the FileWatcher and periodically ensures DNS clients
+// are using the current ca bundle until a message is received on the stop
+// or error channels.
+func (m *Provider) startWatcher(stop <-chan struct{}) error {
+	go wait.Until(func() {
+		if err := m.ensureDNSClientsTLSTransport(); err != nil {
+			log.Error(err, "failed to ensure dns client tls transport")
+		}
+	}, 1*time.Minute, stop)
+
+	errChan := make(chan error)
+	go func() {
+		errChan <- m.fileWatcher.Start(stop)
+	}()
+
+	// Wait for the watcher to exit or an explicit stop.
+	select {
+	case <-stop:
+		return nil
+	case err := <-errChan:
+		return err
+	}
+}
+
+// ensureDNSClientsTLSTransport compares the watched ca bundle with the
+// current ca bundle of FileWatcher and updates DNS clients with the current
+// ca bundle if the two are not equal.
+func (m *Provider) ensureDNSClientsTLSTransport() error {
+	equal, caBundle, err := m.caBundlesEqual()
+	if err != nil {
+		return err
+	}
+	if equal {
+		return nil
+	}
+
+	certs, err := certsutil.CertsFromPEM(caBundle)
+	if err != nil {
+		return err
+	}
+
+	m.route53.Client.Config.HTTPClient.Transport = certsutil.MakeTLSTransport(certs)
+	m.elb.Client.Config.HTTPClient.Transport = certsutil.MakeTLSTransport(certs)
+	m.tags.Client.Config.HTTPClient.Transport = certsutil.MakeTLSTransport(certs)
+
+	return nil
+}
+
+// caBundlesEqual compares the watched ca bundle with the current
+// ca bundle of FileWatcher, returning false and the current ca
+// bundle if the two are not equal.
+func (m *Provider) caBundlesEqual() (bool, []byte, error) {
+	watchedCAs, err := ioutil.ReadFile(m.fileWatcher.GetFile())
```

Comment:
This is expensive.  Can we just compare `Stat` mtimes, like I floated in golang/go#35887?  I tested this locally, bumping the ConfigMap content and seeing the mtime change (although you need `Stat` to walk the symlink to see the real mtime in the linked file).  On the other hand, reading the certs once a minute shouldn't be that bad, even if the reads are expensive ;).

Reply:
And I guess more fundamentally, if we're going to read the whole file off the disk, it seems like we might as well update the CAs without bothering to check if there were any changes.  It's not obvious to me that a bytewise comparison in memory is all that much cheaper than just parsing out the certs into Go CAs.

Reply:
Another way might be to forgo fs notifications entirely and wrap access to the underlying file with a struct that implements a TTL for refresh. So like, calling CertFile.Get() would return cached data until its internal TTL (e.g. 30s, 1min) expires, at which point it will lock/refresh/re-cache/reset TTL. Such an approach wouldn't require any goroutines at all.

Reply:
Performance considerations seem pretty inconsequential in this context — we're talking about reading a secret mount, which I believe is memory mapped? We could unconditionally re-read the contents every 15 seconds or something without any diffing at all and I'm not sure it would have any consequential impact on the platform